src/main/java/graphql/execution/ExecutionContext.java

```diff
@@ -80,7 +84,17 @@ public class ExecutionContext {
         this.errors.set(builder.errors);
         this.localContext = builder.localContext;
         this.executionInput = builder.executionInput;
-        queryTree = FpKit.interThreadMemoize(() -> ExecutableNormalizedOperationFactory.createExecutableNormalizedOperation(graphQLSchema, operationDefinition, fragmentsByName, coercedVariables));
+        this.queryTree = FpKit.interThreadMemoize(() -> ExecutableNormalizedOperationFactory.createExecutableNormalizedOperation(graphQLSchema, operationDefinition, fragmentsByName, coercedVariables));
+        this.rootDataFetchingFieldSelectionSet = FpKit.intraThreadMemoize(() -> collectRootDataFetchingFieldSelectionSets(queryTree, builder.graphQLSchema));
```

Comment:
Intra-thread as the creating `DataFetchingFieldSelectionSet` isn't an expensive operation in itself to warrant synchronized access. But that might be a wrong conclusion.